service/history/ndc/hsm_state_replicator.go

```diff
@@ -183,29 +197,108 @@ func (r *HSMStateReplicatorImpl) syncHSMNode(
 		incomingNodePath := incomingNode.Path()
 		currentNode, err := currentHSM.Child(incomingNodePath)
 		if err != nil {
-			// 1. Already done history resend if needed before,
-			// and node creation today always associated with an event
-			// 2. Node deletion is not supported right now.
-			// Based on 1 and 2, node should always be found here.
+			// In the old code:
+			//   Node creation is always associated with an event.
+			// Given that events should have been replicated first, we shouldn't miss nodes here.
+			// If we do, it's unexpected, so return the error.
 			return err
 		}
 
-		if shouldSyncNode, err := r.shouldSyncNode(currentNode, incomingNode); err != nil || !shouldSyncNode {
+		shouldSyncNode, err := r.shouldSyncNode(currentNode, incomingNode)
+		if err != nil {
 			if err != nil && errors.Is(err, hsm.ErrInitialTransitionMismatch) {
+				// skip syncing this node if initial transitions mismatch
```

Comment:
Do we assume node path will never be reused? otherwise I don't think we can skip here.

Reply:
We can assume this today but I'd explain that we make this assumption in the code.